buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java

```diff
@@ -0,0 +1,141 @@
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.BuildListener;
+import org.gradle.BuildResult;
+import org.gradle.api.GradleException;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.invocation.Gradle;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * A plugin to handle reaping external services spawned by a build if Gradle dies.
+ */
+public class ReaperPlugin implements Plugin<Project> {
+
+    private Project project;
+    private Path baseDirectory;
+    private volatile Process reaperProcess;
+
+    @Override
+    public void apply(Project project) {
+        this.project = project;
+        this.baseDirectory = project.getRootDir().toPath().resolve(".gradle")
+            .resolve("reaper").resolve("build-" + ProcessHandle.current().pid());
+        project.getGradle().addListener(new BuildListener() {
+            @Override
+            public void buildStarted(Gradle gradle) {}
+
+            @Override
+            public void settingsEvaluated(Settings settings) {}
+
+            @Override
+            public void projectsLoaded(Gradle gradle) {}
+
+            @Override
+            public void projectsEvaluated(Gradle gradle) {}
+
+            @Override
+            public void buildFinished(BuildResult result) {
+                if (reaperProcess != null) {
+                    try {
+                        reaperProcess.getOutputStream().close();
+                        project.getLogger().info("Waiting for reaper to exit normally");
+                        if (reaperProcess.waitFor() != 0) {
+                            throw new GradleException("Reaper process failed. Check log at "
+                                + baseDirectory.resolve("error.log") + " for details");
+                        }
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+
+                }
+            }
+        });
+    }
+
+    /**
+     * Register a pid that will be killed by the reaper.
+     */
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)}).supply();
+        registerCommand(serviceId, killPidCommand);
+    }
+
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
+        ensureReaperStarted();
+
+        try {
+            Files.writeString(baseDirectory.resolve(serviceId), String.join(" ", command));
```

Comment:
Should we be including a newline here somewhere?

Reply:
No need. While the reaper does handle multiple lines, we don't actually use that. It will end up reading just the entire content into a single line from the file.